Update mac jhbuild instructions

Technius opened the pull request:
This PR improves the macOS jhbuild instructions and the mac build readme.
* Move homebrew section to top of the readme
* Add instructions on how to set up jhbuild correctly
* Create a custom jhbuild moduleset for installing 3rd party deps instead of using custom bash scripts
* Add lua package build (not built previous; now it's in the moduleset)

A completely fresh build works locally on my Mac M1 Mini running macOS 12.4 Monterey.

Fixes #3949. Fixes #4222.

Technius commented:
I think this might not be compatible with the release pipelines at the moment (due to the binary blob & the required `UNDER_JHBUILD` check).

rolandlo commented on mac-setup/xournalpp.modules:
```diff
@@ -0,0 +1,85 @@
+
+  <cmake id="libzip" cmakeargs="-DENABLE_ZSTD=OFF -DCMAKE_INSTALL_NAME_DIR='&#36;&lt;INSTALL_PREFIX&gt;/lib'">
+    <branch repo="libzip"
+            version="1.8.0"
+            module="libzip-1.8.0.tar.gz"/>
```
The latest version is 1.9.2.
```suggestion
            version="1.9.2"
            module="libzip-1.9.2.tar.gz"/>
```

rolandlo commented on mac-setup/xournalpp.modules:
```diff
@@ -0,0 +1,85 @@
+             uninstall-before-install="yes"
+             makeinstallargs="INSTALL_TOP='$(DESTDIR)/${prefix}' install"
+  >
+    <branch repo="lua"
+            version="5.4.4"
+            module="lua-5.4.4.tar.gz"
+            checkoutdir="lua-5.4.4"/>
+  </autotools>
```
Maybe add a comment, that this requires `cmake >= 3.18`.

Technius replied:
gtk-osx pulls in CMake 3.20, last time I checked.

Technius replied:
Also, I'm confused why we need CMake for lua here. Lua uses a custom Makefile.

rolandlo replied:
Not for building Lua, but for using it to build Xournal++ via `cmake`
https://github.com/xournalpp/xournalpp/blob/024c3419f1194adefa72e37e7fcdf408065d5f2b/CMakeLists.txt#L126-L136

The question is if we have `cmake >=3.18` in Azure pipelines.

rolandlo replied:
Ok, when it is in the moduleset, it will be installed in `$HOME/gtk` and it can be removed from `build_mac.yml` after updating the Gtk Mac blob. 
https://github.com/xournalpp/xournalpp/blob/024c3419f1194adefa72e37e7fcdf408065d5f2b/azure-pipelines/steps/build_mac.yml#L37-L45

rolandlo commented on readme/MacBuild.md:
````diff
@@ -1,34 +1,175 @@
+cmake --install .
+cd ..
+
+# build application
+cd mac-setup
+jhbuild run bash ./build-app.sh "$HOME"/gtk
+```
````
Maybe express that building an app bundle is optional and that at this stage Xournal++ can be run via 
`jhbuild run xournalpp`

Technius replied:
`jhbuild run xournalpp` does not work, so I changed this to mention that the app bundle should be installed first.

rolandlo replied:
`jhbuild run xournalpp` works after `cmake --install .` There is no need to run `build-app.sh` (creating Xournal++.app) for testing it out from the command line.

Technius replied:
It doesn't work for me on the special build user, since my build user does not have GUI access.

rolandlo commented on readme/MacBuild.md:
`````diff
@@ -38,33 +179,10 @@ Follow these steps to resolve the issue:
-#### Potential Errors
-##### Unknown module (macOS Mojave 10.14)
+#### Unknown module error (macOS Mojave 10.14)
 
 If there is an error like:
-````bash
+````sh
`````
Should we really still include comments about macOS Mojave?

Technius replied:
This is copied over from the old readme. Not sure if this issue still occurs...

Technius replied:
I deleted this since it is likely that this issue has been resolved in jhbuild a long time ago.

rolandlo commented on readme/MacBuild.md:
`````diff
@@ -80,71 +198,34 @@ Follow these steps to resolve the issue:
-````bash
-./build-app.sh $HOME/gtk
-````
+# change this to point to your own user directory
+modulesets_dir = "/Users/jhbuilduser/gtk-osx/modulesets-stable"
+```
`````
If Xournal++ is built for the Gtk Mac blob, `jhbuilduser` should be `xournal-dev`. Maybe this could be said at some place, since it might save some time for a core team member, who doesn't know that.

Technius replied:
Replaced with something that should work with any username.

rolandlo reviewed (commented):
I have some comments right now, before I try it out.

One issue that I encounter(ed) is that the `jhbuildrc-gtk-osx` python script from the `gtk-osx` repo doesn't detect my `xcode` version (I only have the command line tools installed) and aborts, 

https://gitlab.gnome.org/GNOME/gtk-osx/-/blob/master/jhbuildrc-gtk-osx#L137-162

I make it return `11.5` (at the beginning of the `else`-block) to make it work. Have you encountered that problem as well?

rolandlo commented on readme/MacBuild.md:
```diff
@@ -1,34 +1,175 @@
+If you encounter a yet unknown error, feel free to submit a pull request to add it to this set of instructions.
+
+
+### Step 1: Install jhbuild
```
Step 1 should be to clone the `xournalpp` github repository and `cd` into `mac-setup` folder. This step is completely missing here.

Technius replied:
That doesn't make sense in this context; we can't even build xournalpp without having jhbuild set up first. I have added a sentence clarifying that the reader needs to be logged in as the jhbuild user.

Technius replied:
I inserted a cloning step between the "build dependencies from gtk-osx" and "build dependencies not in gtk-osx"

rolandlo commented on readme/MacBuild.md:
```diff
@@ -1,34 +1,175 @@
+bash gtk-osx-setup.sh
+
+# add jhbuild to path
+echo 'export PATH="$HOME"/.new_local/bin:"$PATH"' >> ~/.zshrc
```
After this, the console should be reopened, since otherwise the old `PATH` is still in use.

Technius replied:
The user could also run that export command directly. I've added methods below this line.

rolandlo commented on readme/MacBuild.md:
```diff
@@ -1,34 +1,175 @@
+jhbuild bootstrap-gtk-osx
+
+# if the build fails because pygments is not found, try installing it manually:
+jhbuild build pygments
```
Just a remark: For me this was not needed (with a clean `Users/xournal-dev` folder). Previously I did `pip3 install pygments` instead to solve this.

Technius replied:
I can't reproduce this issue on a clean build anymore.

Technius replied:
But I think I'll leave this here just in case (maybe we can move it to the FAQ in the future?)

rolandlo commented on readme/MacBuild.md:
````diff
@@ -1,34 +1,175 @@
+To build them, run the following command in the `mac-setup` folder:
+
+```sh
+jhbuild -m xournalpp.modules build meta-xournalpp-deps
````
Steps 1 and 2 worked fine (modulo the minor issues I mentioned above). 

That command does not work though, since `use_local_modulesets` is `False` by default. So `jhbuild` searches the module online. Probably you cloned the `gtk-osx` repository and used only local modulesets.

Technius replied:
Given that the developer needs to already create a separate user for building the application, it might be okay to require `use_local_modulesets` then.

rolandlo replied:
> Given that the developer needs to already create a separate user for building the application, it might be okay to require `use_local_modulesets` then.

I'm fine with that. Just don't forget to include the step to clone/update the `gtk-osx` repository and modify `~/.config/jhbuildrc-custom` step then.

rolandlo reviewed (changes requested):
I have added some more comments. Further changes are needed.

rolandlo commented on mac-setup/xournalpp.modules:
```diff
@@ -0,0 +1,85 @@
+  </cmake>
+
+  <autotools id="portaudio" autogenargs="--enable-cxx --disable-mac-universal">
+    <branch repo="github-tarball"
```
I think you wanted this (the Github repo doesn't have that file, it seems)
```suggestion
    <branch repo="portaudio"
```

Technius commented:
Sorry, I haven't had time to work on this recently, and I'll be away from my Mac device in the two weeks. I'll update the readme but I won't be able to test the build until afterwards.

rolandlo commented on readme/MacBuild.md:
````diff
@@ -1,141 +1,264 @@
-
-### Automated step
+You also run the application from the `build` folder without installing:
 
 ```sh
-./complete-build.sh $HOME/gtk
+cd build
+jhbuild run ./xournalpp
````
I don't think that works as of now (c.f. #4153). What you can do is run the installed version using `jhbuild run xournalpp` without creating the .app-bundle.

Technius replied:
Actually, I can't get this to run with the special build user. I'll need to see if this works with the binary blob.

Technius replied:
It doesn't work with the binary blob, unfortunately.

rolandlo commented on readme/MacBuild.md:
```diff
@@ -1,141 +1,264 @@
+If you are using Python 3.10 (check with `jhbuild run python --version`), there
+is a bug in `libxml2` that causes it to segfault when imported by itstool.
+
+To work around this problem, you will need to reinstall gtk-osx using the script
+above.
```
It's not clear what "the script above" refers to.

rolandlo commented on mac-setup/build-app.sh:
```diff
@@ -1,10 +1,18 @@
+#
+# This script should be run from a jhbuild environment
+
+if [ -z "$UNDER_JHBUILD" ]; then
+  echo 'Error: this command needs to be run within jhbuild.'
+  exit 1
+fi
```
We still use `build-app.sh` on Azure Pipelines, where we have no `jhbuild`.

Technius replied:
That's a good point.

It might be good to also add instructions on how to create a binary blob. This would be convenient for building the dependencies on a special jhbuild user account, and then using the regular user account to build the final xournalpp executable & bundle.

rolandlo replied:
Fortunately that's quite simple nowadays: Just download the script https://raw.githubusercontent.com/xournalpp/xournalpp-pipeline-dependencies/base/gtk/package-gtk-bin.sh and run it via
```term
./package-gtk-bin.sh "$HOME"/gtk
```
Extract the split up archive via
```term
cat gtk-bin.tar.gz.* | tar -xz
```

Technius replied:
After creating a binary blob with a special jhbuild user account and trying to use it on a regular user account, I think this is too difficult. There's too many paths that can get messed up by moving the gtk prefix, so we'll have to assume that the app is built with the special jhbuild user account.

I think we can at least document how the blob build works though.

Technius replied:
I changed this check so that it only prints out a non-fatal warning instead of quitting.

rolandlo reviewed (changes requested):
This looks much better already. Some issues remain though (see comments).

Also you might want to remove `build-sndfile.sh` and `build-libzip.sh`.

hyperupcall commented:
Hey - I made a PR at #4536 to improve some of the shell scripts, but it was brought to my attention that much of my changes would be removed / would clash with the changeset here.

So I closed that PR and checked over the changes here for similarities, and if you don't mind, I made (will make) a few suggestions

hyperupcall commented on mac-setup/build-app.sh:
```diff
@@ -1,10 +1,18 @@
 #!/bin/bash
```
Would it be possible to use a consistent shebang: `#!/usr/bin/env bash`? I noticed the old one was already replaced in `mac-setup/complete-build.sh`

hyperupcall commented on mac-setup/build-app.sh:
```diff
@@ -16,34 +24,33 @@ cd "${0%/*}" || exit
-make install
-cd ..
+# add gtk-mac-bundler to PATH
+export PATH="$PATH":"$HOME"/.local/bin
```
`PATH` is already exported, so the `export` is redundant

Technius replied:
Doesn't hurt to be explicit here.

hyperupcall commented on mac-setup/build-app.sh:
```diff
@@ -16,34 +24,33 @@ cd "${0%/*}" || exit
+# add gtk-mac-bundler to PATH
+export PATH="$PATH":"$HOME"/.local/bin
+
+GTK_MAC_BUNDLER=$(which gtk-mac-bundler)
```
Prefer `command -v` (or the usage of any builtin) over `which`

Technius replied:
This doesn't really matter since this script is assumed to run on macOS, and the `which` executable always exists. `command -v` is not portable across shells.

hyperupcall replied:
It is `which` that's generally not portable across shells. There are dozens of different implementations, each with different sets of flags and peculiarities; sometimes it's a builtin, other times it's an executable - and it's not in any known standard I know of. `command[ -v]` and its behavior has been clearly specified as a shell builtin in POSIX for decades

I'd say none of my suggestions here "really matter" in the sense that they significantly change the behavior of the script. I suppose I shouldn't have suggested them then, and maybe my bias and personal style bleed through these recommendations. Should probably resign to the fact that these things sneak in

hyperupcall replied:
I tried to resolve my comments, but it looks like they've already been resolved. Happy to see this improved building experience for Mac

rolandlo commented on mac-setup/xournalpp.modules:
```diff
@@ -0,0 +1,85 @@
+  <!-- For poppler-glib -->
+  <cmake id="poppler" cmakeargs="-DENABLE_QT5=OFF -DENABLE_QT6=OFF -DENABLE_BOOST=OFF -DENABLE_CPP=OFF -DENABLE_LIBOPENJPEG=none -DENABLE_CMS=none -DCMAKE_INSTALL_NAME_DIR='&#36;&lt;INSTALL_PREFIX&gt;/lib'">
+    <branch repo="poppler"
+            version="22.10.0"
+            module="poppler-22.10.0.tar.xz"/>
```
Poppler has recently been updated.
```suggestion
            version="22.12.0"
            module="poppler-22.12.0.tar.xz"/>
```

hyperupcall commented:
Would it also be possible to move the the Autotools poppler-specific stuff from `.gitignore` to `mac-setup/.gitignore`? It doesn't seem necessary to leave it at the root anymore

hyperupcall commented on mac-setup/complete-build.sh:
```diff
@@ -1,20 +1,25 @@
-export PATH="$HOME/.local/bin:$1/inst/bin:$PATH"
+root_dir=$(readlink -f "$(dirname "$0")"/..)
+prefix_parent="$1"
+[ ! -d "$prefix_parent" ] && exit "$prefix_parent doesn't exist."
```
This just caught my eye, but Bash only allows the single argument of `exit` to be an integer representing the exit code, not a string

Technius replied:
Thanks for the catch. I've been writing too much python lately, where `sys.exit("message")` does exactly this...

Technius commented:
Finally had some time to play around with this more.
* I updated my macOS devices to Ventura (which forced me to remove my gtk installation), so I had a chance to try the instructions again. I ran into more issues, which I've documented in the instructions.
* I added a script for building a `.dmg`, which conveniently allows users to drag the `.app` into the `Applications` folder. It's not called in `build-app.sh`, since this will require some CI script changes.
* `complete-build.sh` now dynamically detects whether it's being run under jhbuild, so it doesn't try to run under jhbuild recursively.

Still working on addressing some of the other feedback...

Technius commented:
/azp run create-installers

LittleHuba commented:
The requested installers were generated successfully. The installers are available at:

- [Windows](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjU1MjAvYXJ0aWZhY3ROYW1lL3dpbmRvd3M1/content?format=zip)
- [macOS](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjU1MjAvYXJ0aWZhY3ROYW1lL21hY09T0/content?format=zip)
- [Debian](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjU1MjAvYXJ0aWZhY3ROYW1lL2xpbnV4LWRlYmlhbi1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 18.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjU1MjAvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0xOC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 20.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjU1MjAvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 22.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjU1MjAvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMi4wNC1wYWNrYWdlcw2/content?format=zip)

Please be aware that these links are only valid for a limited time (~30 days).

rolandlo reviewed (commented):
It looks mostly good now. Still I have some comments that need to be addressed.

rolandlo commented:
By the way, `itstool` just got removed from the `gtk-osx` project, see https://gitlab.gnome.org/GNOME/gtk-osx/-/commit/013ef29a409c4c76d2c146626dca877bc7c9b602. So the workaround with fixing the python version shouldn't be needed anymore.

Technius commented on mac-setup/complete-build.sh:
```diff
@@ -1,20 +1,35 @@
+    ENSURE_JHBUILD="jhbuild run"
+fi
+
+root_dir=$(readlink -f "$(dirname "$0")"/..)
```
Maybe realpath would be better here.

Technius replied:
Replaced with a Python pathlib solution instead

Technius commented:
Looks like the bash Ventura issue was fixed as well.

Technius commented on mac-setup/complete-build.sh:
```diff
@@ -1,20 +1,35 @@
+if [ -z "$UNDER_JHBUILD" ]; then
+    echo "Not running jhbuild environment, prepending jhbuild to every command"
+    if ! which jhbuild > /dev/null; then
+        echo "Error: jhbuild not found on PATH!" && exit 1
```
Ok, maybe we can make this a warning instead, since the script already has `-e` set and pipefail enabled.

Technius commented:
Side note: I'm currently experimenting with a CI script running on [Cirrus CI](https://cirrus-ci.com). This is out of scope of this PR, but if all goes well, we should be able to automatically build the binary blob & get automated builds for both x86 & ARM macs.

Technius commented:
I squashed the fixups from the previous round of reviews and incorporated the new changes into the latest fixups. I also improved parts of the build process based on my experience (successfully) building the application bundle with Cirrus CI.

rolandlo reviewed (approved):
LGTM!

Technius commented:
I added some additional information to the README that I forgot to note down.

Will squash the fixups and merge in 48 hours if there are no objections.

Technius commented:
/azp run create-installers

LittleHuba commented:
The requested installers were generated successfully. The installers are available at:

- [Windows](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjY1MDAvYXJ0aWZhY3ROYW1lL3dpbmRvd3M1/content?format=zip)
- [macOS](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjY1MDAvYXJ0aWZhY3ROYW1lL21hY09T0/content?format=zip)
- [Debian](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjY1MDAvYXJ0aWZhY3ROYW1lL2xpbnV4LWRlYmlhbi1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 20.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjY1MDAvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMC4wNC1wYWNrYWdlcw2/content?format=zip)
- [Ubuntu 22.04](https://artprodsu6weu.artifacts.visualstudio.com/A4bb312fe-95d7-456e-96fb-7669bb27c94b/ce85de88-4b71-44a0-8a57-2707c0b9a6e7/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL3hvdXJuYWxwcC9wcm9qZWN0SWQvY2U4NWRlODgtNGI3MS00NGEwLThhNTctMjcwN2MwYjlhNmU3L2J1aWxkSWQvMjY1MDAvYXJ0aWZhY3ROYW1lL2xpbnV4LXVidW50dS0yMi4wNC1wYWNrYWdlcw2/content?format=zip)

Please be aware that these links are only valid for a limited time (~30 days).
